Propagate index errors in field_caps

cbuescher opened the pull request:
Currently we don't report any exceptions occuring during `field_caps` requestrs back to the user.
This PR adds a new `failure` section to the response which contains exceptions per index, keyed by
the index name the error occured in. In addition the response contains another field `failed_indices`
which reports the number of indices that threw an exception. If all of the requested indices failed,
the whole request fails with a 500 error code, otherwise the request succeeds and it is up to the caller
to check for potential errors in the response body.

Closes #68994

Commit:
Propagate errors in field_caps

cbuescher commented:
This is an example of the response with failures with the changes in this PR:
```
{
	"indices": [...],
	"fields": {
		...
	},
	"failed_indices": 1,
	"failures": {
		"errorindex": {
			"error": {
				"root_cause": [{
					"type": "illegal_argument_exception",
					"reason": "test"
				}],
				"type": "illegal_argument_exception",
				"reason": "test"
			}
		}
	}
}
```

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

jimczi commented on server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:
```diff
@@ -25,18 +26,28 @@
     private final String indexName;
     private final Map<String, IndexFieldCapabilities> responseMap;
     private final boolean canMatch;
+    private final Exception exception;
```
I don't think the exception should be part of the index response. It's already reported in the coordinator node so we don't need this hybrid approach.  `FieldCapabilitiesIndexResponse` should represent a successful response.

jimczi commented on server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:
```diff
@@ -77,47 +77,43 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        indexResponses.add(new FieldCapabilitiesIndexResponse(index, e));
```
We can gather the exception in a synchronized map, no need to build fake `FieldCapabilitiesIndexResponse`.

jimczi commented on server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:
```diff
@@ -135,27 +131,52 @@ private void countDown() {
+            remoteClusterClient.fieldCaps(remoteRequest, ActionListener.wrap(response -> {
+                List<FieldCapabilitiesIndexResponse> remotes = new ArrayList<>();
+                for (FieldCapabilitiesIndexResponse resp : response.getIndexResponses()) {
+                    if (resp.getException() == null) {
```
Same here, we can simply add the failures map of the remote response to the local one.

jimczi reviewed (commented):
The approach looks good to me. I left one comment regarding the reporting of the exception.

cbuescher commented:
@jimczi thanks for the review, I will rework the part around how to transport back the exceptions.

Commit:
Merge branch 'master' into fix-68994

Commit:
Adress comments

cbuescher commented:
@jimczi I pushed and update reworking how the exceptions get transported back.

jimczi commented on server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:
```diff
@@ -135,25 +132,36 @@ private void countDown() {
+                countDown.run();
+            }, e -> {
+                // individual index failures should show up as remote responses containing an exception
+                countDown.run();
```
This case is not handled automatically. We need to add all the indices from that cluster in the index failures map. Since all indices on that cluster failed, maybe that the key in the map could be the cluster name instead ?

cbuescher replied:
Individual index failures on a remote should be handled (there is also a test for it now that should fail otherwise). So you mean this should cover general remote failures? How could this happen? Also asking since I cannot thing around a way of testing this atm...

Commit:
Iter: group similar exceptions

Commit:
Merge branch 'master' into HEAD

Commit:
iter

jimczi commented on server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesResponse.java:
```diff
@@ -190,4 +230,19 @@ public int hashCode() {
+    @Override
+    public RestStatus status() {
+        RestStatus status = RestStatus.OK;
+        if (indices.length == 0 && failures.size() > 0) {
```
Do we need this logic now that we throw an exception if all indices fail ? We could even assert in the ctr that failures  must be empty if indices is empty.

jimczi reviewed (commented):
Looks good to me. Could you add a test a yml test for ccs as well ?

Commit:
Merge branch 'master' into fix-68994

Commit:
Remove rest status logic

Commit:
Simplify failure collection in one place

Commit:
Add multi cluster yml test

Commit:
checkstyle

Commit:
cleanup

cbuescher commented:
@jimczi thanks, I pushed changes that remove the rest status logic you mentioned, reworked the failure collection a bit to be all on a central place and added the test you asked for.

Commit:
Merge branch 'master' into fix-68994

jimczi reviewed (approved):
LGTM, thanks @cbuescher !
